### ISSUE TITLE
triggered execution (file watch) should print debug log of the changed files

### ISSUE THREAD

**Issue by eed3si9n** [sbt/sbt#3869]
## steps

1. `~` causes infinite compilation loop.

## problem

We'd like to know which file got changed, but we can't because `WatchState#sources` is private.

## expectation

`~` should print changed files in debug level.

**Comment by eed3si9n** [comment]
Note to a potential contributor: 

Probably add stuff here:

https://github.com/sbt/sbt/blob/v1.1.0/main-command/src/main/scala/sbt/Watched.scala#L100-L101

```scala
    if (watchState.count > 0)
      printIfDefined(watched watchingMessage watchState)
```

**Comment by dwijnand** [comment]
Not it'll be in this if block: https://github.com/sbt/io/blob/v1.1.3/io/src/main/scala/sbt/internal/io/SourceModificationWatch.scala#L58

**Comment by nafg** [comment]
Why not just make `WatchState#sources` not private?

**Comment by eatkins** [comment]
That doesn't fix the problem. `WatchState#sources` tells you what sources _could_ have triggered the rebuild, but not which actually did. At any rate, the logs that show explicitly which file triggered the build are now present, but due to #4097, they are not yet actually visible.

**Comment by nafg** [comment]
Is that somehow different than the relevant settings? If not, what purpose
does it serve?

**Comment by nafg** [comment]
Also, if it is logged in the debug log stream together with a million other things, it's not that helpful. If there was just a field on WatchState or something else I can access in the `triggeredMessage` setting, I could just get it to print there without affecting anything else.

**Comment by eatkins** [comment]
> Is that somehow different than the relevant settings? If not, what purpose
does it serve?

I don't understand the question. Would you elaborate?

> Also, if it is logged in the debug log stream together with a million other things, it's not that helpful. If there was just a field on WatchState or something else I can access in the triggeredMessage setting, I could just get it to print there without affecting anything else.

The purpose of the logs are for debugging. They weren't added with the idea of building any additional feature around them. Yes, the log output is noisy in debug mode, but ideally the feature works correctly the vast majority of the time and the noise is irrelevant.

I like the idea of adding an additional Path parameter to triggeredMessage, but that isn't straightforward due to binary compatibility. At any rate, it's a feature request. I'm not going to add it to my queue at the moment, but it would be a reasonable project for a new contributor if we scope out how to address the binary compatibility issue.

**Comment by nafg** [comment]
I mean, is the value of WatchState#source different than the value of the watchSources setting

**Comment by eatkins** [comment]
This functionality was added in: https://github.com/sbt/sbt/pull/4335
To separate the watch debug messages from the regular debug messages, the user can override watchLogger. Note that this will likely only be available in sbt > 1.3.
